File: amino/backends/postgres.py
"""Postgres compile target. Emits psycopg-style positional ``%s`` parameters."""

from typing import Any

from .base import ParamSink, SQLBackend


class _PositionalParams(ParamSink):
    def __init__(self) -> None:
        self._values: list[Any] = []

    def add(self, value: Any, base_type: str) -> str:
        self._values.append(value)
        return "%s"

    def result(self) -> list[Any]:
        return self._values


class PostgresBackend(SQLBackend):
    """Compile an expression to a Postgres ``WHERE`` predicate.

    ``Query.params`` is a list suitable for ``cursor.execute(sql, params)``
    with psycopg. List membership renders as ``col = ANY(%s)`` so a Python
    list binds as a single array parameter.
    """

    def quote_ident(self, name: str) -> str:
        return '"' + name.replace('"', '""') + '"'

    def new_params(self) -> ParamSink:
        return _PositionalParams()

    def render_in(self, column: str, placeholder: str) -> str:
        return f"{column} = ANY({placeholder})"

    def render_contains(self, haystack: str, needle: str) -> str:
        return f"position({needle} IN {haystack}) > 0"

File: go/backend/postgres.go
package backend

import (
	"strings"

	amino "github.com/raiderrobert/amino/go"
)

// Postgres compiles to a predicate with positional $1, $2 placeholders, the
// style pgx and database/sql's Postgres drivers expect. List membership
// renders as col = ANY($n) so a Go slice binds as one array parameter.
type Postgres struct {
	// Columns maps a field path to a SQL fragment rendered verbatim.
	Columns map[string]string
}

// Compile renders expr for Postgres.
func (p Postgres) Compile(expr *amino.Expression) (Query, error) {
	return Compile(pgDialect{}, expr, p.Columns)
}

type pgDialect struct{}

type pgParams struct{ values []any }

func (s *pgParams) Add(value any, _ string) string {
	s.values = append(s.values, value)
	return "$" + itoa(len(s.values))
}

func (s *pgParams) Result() any {
	if s.values == nil {
		return []any{}
	}
	return s.values
}

func (pgDialect) Name() string { return "Postgres" }
func (pgDialect) QuoteIdent(name string) string {
	return `"` + strings.ReplaceAll(name, `"`, `""`) + `"`
}
func (pgDialect) NewParams() ParamSink              { return &pgParams{} }
func (pgDialect) RenderIn(col, ph string) string    { return col + " = ANY(" + ph + ")" }
func (pgDialect) RenderNotIn(col, ph string) string { return "NOT (" + col + " = ANY(" + ph + "))" }
func (pgDialect) RenderContains(hay, needle string) string {
	return "position(" + needle + " IN " + hay + ") > 0"
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	var b [20]byte
	i := len(b)
	for n > 0 {
		i--
		b[i] = byte('0' + n%10)
		n /= 10
	}
	return string(b[i:])
}
